common/src/main/java/tachyon/client/BlockMasterClient.java

```diff
@@ -0,0 +1,242 @@
+/*
+ * Licensed to the University of California, Berkeley under one or more contributor license
+ * agreements. See the NOTICE file distributed with this work for additional information regarding
+ * copyright ownership. The ASF licenses this file to You under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License. You may obtain a
+ * copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package tachyon.client;
+
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ExecutorService;
+
+import org.apache.thrift.TException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import tachyon.Constants;
+import tachyon.MasterClientBase;
+import tachyon.conf.TachyonConf;
+import tachyon.thrift.BlockInfo;
+import tachyon.thrift.BlockMasterService;
+import tachyon.thrift.Command;
+import tachyon.thrift.NetAddress;
+import tachyon.thrift.WorkerInfo;
+
+/**
+ * A wrapper for the thrift client to interact with the block master, used by tachyon clients.
+ *
+ * Since thrift clients are not thread safe, this class is a wrapper to provide thread safety.
+ */
+// TODO: better deal with exceptions.
+public final class BlockMasterClient extends MasterClientBase {
+  private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
+
+  private BlockMasterService.Client mClient = null;
+
+  public BlockMasterClient(InetSocketAddress masterAddress, ExecutorService executorService,
+      TachyonConf tachyonConf) {
+    super(masterAddress, executorService, tachyonConf);
+  }
+
+  @Override
+  protected String getServiceName() {
+    return Constants.BLOCK_MASTER_SERVICE_NAME;
+  }
+
+  @Override
+  protected void afterConnect() {
+    mClient = new BlockMasterService.Client(mProtocol);
+  }
+
+  @Override
+  protected void afterDisconnect() {
+  }
+
+  /**
+   * Get the info of a list of workers.
+   *
+   * @return A list of worker info returned by master
+   * @throws IOException
+   */
+  public synchronized List<WorkerInfo> getWorkerInfoList() throws IOException {
+    while (!mIsClosed) {
+      connect();
+
+      try {
+        return mClient.getWorkerInfoList();
+      } catch (TException e) {
+        LOG.error(e.getMessage(), e);
+        mConnected = false;
+      }
+    }
```

Comment:
I think the infinite loop logic is not a good idea, perhaps we should leave a TODO to guard against this? For example if the network is not configured properly, the client will forever hang.